### PR TITLE
Add email link to set password. See issue #1205.

### PR DESCRIPTION
**Pull request by kiatng**
### Description (*)

When a customer account is created in backed, the password should not be emailed in plaintext. This PR add a new email template, which is modifed from the welcome email, by adding a link to set password.

image

 For existing account, the email template used is the same as the forgot-password email.

This PR does not change existing feature on emailing plaintext password from backend, but it'll show a warning:
image

### Fixed Issues (if relevant)

1. Fixes OpenMage/magento-lts#1205

### Manual testing scenarios (*)

There are 2 scenarios:
#### Creating New Customer
1. In backend, create a new customer with an email address which you own.
2. When you receive the email, click on the link to set the password.
3. Login with the new password.

#### Existing Customer
1. In backend customer page, go to the customer you created above.
2. Click on _Email Link to Set Password_.
3. When you receive the email, click on the link to reset the password.
4. Login with the new password.

### Questions or comments

The landing page of the link is `customer/account/changeforgotten` with page title **RESET PASSWORD**. See [screenshot](https://github.com/OpenMage/magento-lts/issues/1205#issuecomment-690806216). Does it need to be changed?

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/locale/en_US/template/email/account_new_password_link.html [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                    <td>
+                        <a href="{{store url="customer/account/resetpassword/" _query_id=$customer.rp_customer_id _query_token=$customer.rp_token}}"><span>Set Password</span></a>
+                    </td>
+                </tr>
+            </table>
+            <p>To log in when visiting our site just click <a href="{{store url="customer/account/"}}">Login</a> or <a href="{{store url="customer/account/"}}">My Account</a> at the top of every page, and then enter your email address and password.</p>
+            <p class="highlighted-text">
+                Use the following values when prompted to log in:<br/>
+                <strong>Email</strong>: {{var customer.email}}<br/>
+                <strong>Password</strong>: (the password you set)
```
Comment (rjocoleman) [inline app/locale/en_US/template/email/account_new_password_link.html:33]: isn't this line redundant/confusing?

File: app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Newpass.php [hunk 2]
```diff
@@ -49,6 +49,7 @@ public function render(Varien_Data_Form_Element_Abstract $element)
         if ($element->getNote()) {
             $html .= '<p class="note"><span>' . $element->getNote() . '</span></p>';
         }
+        $html .= '<p id="email-passowrd-warning" style="display:none;" class="note"><span>' . Mage::helper('customer')->__('Warning: email cotains password in plaintext will be sent.') . '</span></p>';
```
Comment (rjocoleman) [inline app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Newpass.php:52]: this isn't true because a link is sent instead of a password?

Comment (rjocoleman) [reply app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Newpass.php:52]: There's also a typo: `cotains`

Comment (kiatng) [reply app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Newpass.php:52]: > IMO the automatic email for a manually set password should be removed (if the manual password set feature is retained).

Hmmm... but it's hard to know if a use case exists where it's necessary to send a plaintext password to customer.

Anyway, the objective of this PR is to provide the UI for the backend admin to email a link to set password, especially for customer accounts created in the backend. The automatic email is an existng feature that can be addressed separetely, I don't want to do too many things.

File: app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Adminpass.php [hunk 3]
```diff
@@ -63,6 +63,13 @@ protected function _getScriptHtml(Varien_Data_Form_Element_Abstract $element)
                 $('{$element->getHtmlId()}_container').hide();
                 $('{$element->getHtmlId()}').disable();
             }
+            if ($('email-passowrd-warning')) {
```
Comment (ADDISON74) [inline app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Adminpass.php:66]: Typo: email-**passowrd**-warning

File: app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Adminpass.php [hunk 4]
```diff
@@ -63,6 +63,13 @@ protected function _getScriptHtml(Varien_Data_Form_Element_Abstract $element)
...
+                if (!$('_accountnew_password').getValue() || $('account-send-pass').checked) {
+                    $('email-passowrd-warning').hide();
```
Comment (ADDISON74) [inline app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Adminpass.php:68]: Again: email-**passowrd**-warning

File: app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Adminpass.php [hunk 5]
```diff
@@ -63,6 +63,13 @@ protected function _getScriptHtml(Varien_Data_Form_Element_Abstract $element)
...
+                } else if ($('_accountnew_password').getValue()) {
+                    $('email-passowrd-warning').show();
```
Comment (ADDISON74) [inline app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Adminpass.php:70]: Another one: email-**passowrd**-warning

File: app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Newpass.php [hunk 6]
```diff
@@ -42,6 +42,7 @@ public function render(Varien_Data_Form_Element_Abstract $element)
         if ($element->getNote()) {
             $html .= '<p class="note"><span>' . $element->getNote() . '</span></p>';
         }
+        $html .= '<p id="email-passowrd-warning" style="display:none;" class="note"><span>' . Mage::helper('customer')->__('Warning: an email containing the plaintext password will be sent.') . '</span></p>';
```
Comment (ADDISON74) [inline app/code/core/Mage/Adminhtml/Block/Customer/Edit/Renderer/Newpass.php:45]: email-**passowrd**-warning

**Comment by ADDISON74** [comment]
In "Password Management" section I would make the **Warning** word more visibile using red color. Let's check the format used in other warnings in the Backend. I thing the text is bold too.

**Comment by kiatng** [comment]
> In "Password Management" section I would make the **Warning** word more visibile using red color. Let's check the format used in other warnings in the Backend. I thing the text is bold too.

I copied the format from [Mage_Core.csv](https://github.com/OpenMage/magento-lts/blob/1.9.4.x/app/locale/en_US/Mage_Core.csv#L5)

How it looks now:
image

**Comment by ADDISON74** [comment]
I would come with one more suggestion. Considering that the password is sent in plaintext format, I would insert a paragraph in the email template so that the customer, after receiving the message, should login the account and set another password. The one sent allows him to access the account, but it is not safe because it is sent in plaintext format.

**Comment by kiatng** [comment]
@ADDISON74 indirectly pointed out that there is still one file to change: app/locale/en_US/template/email/password_new.html. This is the only email template where password is sent in plaintext, and it is only sent when the password is set in backend by an admin. When that happens, the customer should be informed of the new password and be advised to set new password ASAP.

File: app/locale/en_US/template/email/password_new.html [hunk 7]
```diff
@@ -16,8 +17,8 @@
     <tr>
         <td class="action-content">
             <h1>{{htmlescape var=$customer.name}},</h1>
-            <p><strong>Password</strong>: (the password you set when creating your account) </p>
-            <p>You can change your password at any time by logging into <a href="{{store url="customer/account/"}}">your account</a>.</p>
+            <p><strong>Your new password is:</strong> {{htmlescape var=$customer.password}}</p>
+            <p>As the password is in plaintext, you should set a new password as soon as possible by logging into <a href="{{store url="customer/account/"}}">your account</a>.</p>
```
Comment (ADDISON74) [inline app/locale/en_US/template/email/password_new.html:21]: Please note the password was sent in the plaintext format. You must set immediately a new password into your <account>.

**Comment by ADDISON74** [comment]
@kiantig - you made a good observation. If the administrator sets the password in the Backend then it is sent to the customer and then it is important that he be notified that the password set by the administrator is not secure. 

In case the customer changes the password, a new message is sent (this is good for security). As far as I remember you made the change so that the password is no longer displayed in that message and that it is mentioned "the one you set".

I haven't checked the templates, but from what I understand the two situations described by me are treated separately. Am I  right?

**Comment by kiatng** [comment]
> In case the customer changes the password, a new message is sent (this is good for security). As far as I remember you made the change so that the password is no longer displayed in that message and that it is mentioned "the one you set".

No, I do not think this is implemented. When a customer changes the password, no email is sent. The changed to not display the password was PR #1021.

**Comment by ADDISON74** [comment]
In this case, I think that the PR just needs to be tested and if there is nothing more to add, we should approve it. Thank you @kiatng for the speed with which you made all the requested changes.

**Comment by fballiano** [comment]
can't test right now cause i'm in the middle of another task, will do asap

**Comment by fballiano** [comment]
@kiatng could you rebase this to main?

**Comment by kiatng** [comment]
I tried to rebase by following this guide: https://gist.github.com/scottyhq/299e4d36018a2f13acfb2528a1553002. But it doesn't seem to work. Is there a guide I can use to rebase?

**Comment by fballiano** [comment]
I follow this guide from @colinmollenhour and it works perfectly: https://github.com/OpenMage/rfcs/pull/10#issuecomment-1496427788

(I always use the "squash" part of the guide)

**Comment by ADDISON74** [comment]
I would close this PR and create a new one because the number of modified files appears to be +5000, being attached from other PRs. I can't test it because there are too many conflicts.

**Comment by kiatng** [comment]
See PR #3262

**Pull request closed without merging**